base/math.jl

```diff
@@ -103,20 +103,23 @@ function evalpoly(x, p::Tuple)
         _evalpoly(x, p)
     end
 end
+evalpoly(x, ::Tuple{}) = zero(one(x)) # dimensionless zero, i.e. 0 * x^0
 
 evalpoly(x, p::AbstractVector) = _evalpoly(x, p)
 
 function _evalpoly(x, p)
     Base.require_one_based_indexing(p)
     N = length(p)
-    ex = p[end]
+    p0 = iszero(N) ? reduce_empty_iter(+, p) : @inbounds p[N]
+    s = oftype(one(x) * p0, p0)
     for i in N-1:-1:1
-        ex = muladd(x, ex, p[i])
+        s = muladd(x, s, @inbounds p[i])
```

Comment:
Since this is a generic function, this is not necessarily safe. It's entirely possible that `muladd(x, y, z)` does some  internal indexing of some array, and the `@inbounds` annotation on the outside would then infect the implementation of `muladd(x, y, z)`. 

An alternative would be to write
```julia
pi = @inbounds p[i]
muladd(x, s, pi)
```